### PR TITLE
Use nonempty list for record fields

### PR DESCRIPTION
**Pull request by smuenzel**
Record labels are currently represented as lists. Empty records, however, are not allowed.
This PR is meant to propose introducing a `Nonempty_list.t` type to enforce the invariant in the parsetree and propagate that to the typechecker. A new nonempty list type is introduced in utils to facilitate this.

If this is a reasonable approach, it could be extended to other invariants, which will end up reducing asserts in the codebase, and prevent future code modifications from introducing violations of the invariants (including in ppxes).

### CODE REVIEW

File: utils/nonempty_list.ml [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+let of_array_exn = function
+  | [||] -> invalid_arg "Nonempty_list.of_array_exn"
+  | ar ->
+      let list = Array.to_list ar in
+      of_list_exn list
+
+let iter f = function
+  | a::l -> f a; List.iter f l
+
+let map f = function
```
Comment (hhugo) [inline utils/nonempty_list.ml:31]:
```ocaml
let map f = function
  | a::l -> let r1 = f a in r1:: List.map f l
```

File: utils/nonempty_list.ml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  | a::l -> let r1 = f a in r1:: List.map f l
+
+let map_to_list f = function
+  | a::l -> let r1 = f a in List.cons r1 (List.map f l)
+
+let[@tail_mod_cons] rec mapi_list i f l =
```
Comment (hhugo) [inline utils/nonempty_list.ml]: Why not use `List.mapi (fun i x -> f (i+1) x)`

File: utils/nonempty_list.ml [hunk 3]
```diff
@@ -0,0 +1,79 @@
+
+type 'a t = (::) of 'a * 'a list
+
+let rec length_aux len (l : _ list) =
+  match l with
+    [] -> len
+  | _::l -> length_aux (len + 1) l
+
+let length (_::l) = length_aux 1 l
```
Comment (hhugo) [inline utils/nonempty_list.ml]: `List.length l + 1`

**Comment by smuenzel** [comment]
@hhugo: I originally had different implementation to try to match the efficiency of `List`, including allocations/depth of function calls. But I see that that is probably overkill now.

File: typing/typecore.ml [hunk 4]
```diff
@@ -3340,12 +3345,12 @@ and type_expect_
             Some {exp with exp_type = ty_exp}, label_definitions
       in
       let num_fields =
-        match lbl_exp_list with [] -> assert false
+        match lbl_exp_list with
```
Comment (hhugo) [inline typing/typecore.ml]: This match could be changed to a `let (_, lbl,_)::_ = lbl_exp_list`.
I'm not sure which is better

File: typing/printtyped.ml [hunk 5]
```diff
@@ -118,6 +118,13 @@ let list i f ppf l =
      List.iter (f (i+1) ppf) l;
      line i ppf "]\n"
 
+let nonempty_list i f ppf l =
```
Comment (hhugo) [inline typing/printtyped.ml]: Maybe better to implement this in term of `list` to avoid duplication

File: typing/printtyp.ml [hunk 6]
```diff
@@ -1468,7 +1468,7 @@ let tree_of_type_decl id decl =
         decl.type_private,
         (rep = Variant_unboxed)
     | Type_record(lbls, rep) ->
-        tree_of_manifest (Otyp_record (List.map tree_of_label lbls)),
+        tree_of_manifest (Otyp_record (Nonempty_list.map_to_list tree_of_label lbls)),
```
Comment (hhugo) [inline typing/printtyp.ml:1471]: Why not propagate the non empty list to otyp_record ?

File: typing/includecore.ml [hunk 7]
```diff
@@ -594,6 +594,8 @@ module Variant_diffing = struct
         | () -> None
       end
     | Types.Cstr_record l1, Types.Cstr_record l2 ->
+        let l1 = Nonempty_list.to_list l1 in
+        let l2 = Nonempty_list.to_list l2 in
         Option.map
           (fun rec_err -> Inline_record rec_err)
           (Record_diffing.compare env ~loc params1 params2 l1 l2)
```
Comment (hhugo) [inline typing/includecore.ml:601]: Should one propagate non empty list to record diffing ?

File: toplevel/genprintval.ml [hunk 8]
```diff
@@ -511,7 +511,7 @@ module Make(O : OBJ)(EVP : EVALPATH with type valu = O.t) = struct
               in
               (lid, v) :: tree_of_fields (pos + 1) remainder
         in
-        Oval_record (tree_of_fields pos lbl_list)
```
Comment (hhugo) [inline toplevel/genprintval.ml:514]: tree_of_fields could be rewritten with mapi. Also should oval_record use non empty list

File: parsing/printast.ml [hunk 9]
```diff
@@ -114,6 +114,13 @@ let list i f ppf l =
      List.iter (f (i+1) ppf) l;
      line i ppf "]\n"
 
+let nonempty_list i f ppf l =
```
Comment (hhugo) [inline parsing/printast.ml]: Rewrite it using `list` above

File: ocamldoc/odoc_sig.ml [hunk 10]
```diff
@@ -970,6 +973,7 @@ module Analyser =
               match types_ext.ext_args with
               | Cstr_tuple l -> Cstr_tuple (List.map (Odoc_env.subst_type env) l)
               | Cstr_record l ->
+                  let l = Nonempty_list.to_list l in
                   let docs = Record.(doc types) pos_end l in
                   Cstr_record (List.map (get_field env docs) l)
             in
```
Comment (hhugo) [inline ocamldoc/odoc_sig.ml:979]: I would expect this to not typecheck

Comment (hhugo) [reply ocamldoc/odoc_sig.ml:979]: Ah we have two constructors with the same name

Comment (smuenzel) [reply ocamldoc/odoc_sig.ml:979]: I believe they are two different types, one defined in `Odoc_info`

File: typing/includecore.ml (hunk 7 above)
Comment (smuenzel) [reply typing/includecore.ml:601]: I think the answer is no. I tried, and it's a little bit of a mess. We would need a new function `Nonempty_list.folding_for_all2`, but we would still end up with many `Nonempty_list.to_list` calls, since we want to use the generic diffing functionality sometimes.

File: lambda/matching.ml [hunk 11]
```diff
@@ -2035,7 +2032,7 @@ let get_pat_args_record num_fields p rem =
   match p with
   | { pat_desc = Tpat_any } -> record_matching_line num_fields [] @ rem
   | { pat_desc = Tpat_record (lbl_pat_list, _) } ->
-      record_matching_line num_fields lbl_pat_list @ rem
+      record_matching_line num_fields (Nonempty_list.to_list lbl_pat_list) @ rem
```
Comment (hhugo) [inline lambda/matching.ml:2035]: record_matching_line could take a Nonempty_list.

File: parsing/ast_invariants.ml [hunk 12]
```diff
@@ -39,9 +38,7 @@ let iterator =
   let super = Ast_iterator.default_iterator in
   let type_declaration self td =
     super.type_declaration self td;
-    let loc = td.ptype_loc in
     match td.ptype_kind with
```
Comment (hhugo) [inline parsing/ast_invariants.ml]: This pattern is now useless

File: parsing/pprintast.ml [hunk 13]
```diff
@@ -1514,7 +1517,7 @@ and type_declaration ctxt f x =
       in pp f "%t%t%a" intro priv variants xs
     | Ptype_abstract -> ()
     | Ptype_record l ->
-        pp f "%t%t@;%a" intro priv (record_declaration ctxt) l
+        pp f "%t%t@;%a" intro priv (record_declaration ctxt) (Nonempty_list.to_list l)
```
Comment (hhugo) [inline parsing/pprintast.ml]: should record_declaration take a Nonempty_list ?

File: lambda/matching.ml (hunk 11 above)
Comment (smuenzel) [reply lambda/matching.ml:2035]: `record_matching_line` takes a `[]` right above

**Comment by hhugo** [comment]
The implementation looks correct.

Regarding the ppx interface/ast change, we should probably check with ppxlib dev (@pitag-ha maybe ?)

**Comment by hhugo** [comment]
It would be useful to list (in the PR description) some other potential usage of Nonempty_list, I think it could help legitimate the introduction of this new data-structure in the compiler

**Comment by gasche** [comment]
I have not looked at the diff in details, but I have mixed feelings about this PR:

- In general I'm not a fan of trying to add too many invariants in the type (cardinality of stuff is "too much" in my book) because it requires adding a lot of conversions in various places, and my feeling is that we (1) easily get outside the comfort zone of the ML type system and (2) don't have clear rules in place for what to move into static types and what to leave out, making these kind of "are we going too far here?" discussion a bit difficult.

- This said, the diff in this PR looks fairly reasonable and actually not so invasive.

- I am not convinced that "empty records are disallowed" is a hard restriction that we want to enforce carefully in the language. I wonder if we should not consider un-enforcing it instead and allow empty records like we allow empty sums. Currently the way to list the labels of a record requires getting hold of a record field, so that does not work, but I don't understand if this is just because the current representation is a ugly hack, or if there is a more fundamental limitation at hand. (I used to remember this more precisely but I forgot, apologies.) I guess the issue is that `{}` in patterns or expressions always require disambiguation. (But not necessarily always type-directed disambiguation. In expressions, the default rule could select the last empty record in scope, and in patterns the default rule could select the last record type in scope. Maybe not doing that and relying on type-directed disambiguation only would be fine.)


So overall:
- I'm not sure about the specific problem that you are trying to fix (disallowing empty records).
- I am skeptical about the general idea/approach of trying to shove all invariants into more complex types.
- But I sort of like your `Nonempty_list` module and wonder where this is going in the compiler codebase.

**Comment by smuenzel** [comment]
> *In general I'm not a fan of trying to add too many invariants in the type

This is definitely a balance,  because currently some invariants come with a very heavy weight enforcement mechanism (e.g. many additional type annotations for gadts). `Nonempty_list`s are a little more lightweight, since they don't really depend on any special mechanisms (other than type-based constructor disambiguation).
One benefit I see with adding invariants is that it makes intentions a lot clearer for the reader -- we currently have `assert`s that "enforce" invariants, but this is not self-documenting, and makes it harder to modify or refactor the codebase. 

> 
> * This said, the diff in this PR looks fairly reasonable and actually not so invasive.
> 
> * I am not convinced that "empty records are disallowed" is a hard restriction that we want to enforce carefully in the language. I wonder if we should not consider un-enforcing it instead and allow empty records like we allow empty sums.

I'm not sure what the semantics of an "empty record" would look like? An empty sum type has no constructors, so cannot be constructed, and a match on it is not possible.
For an empty record, will it be like the `unit` type, a type with a single valid value? In that case, why add another way to do the same thing, when sum types are more flexible (they allow naming of the constructor)?
Additionally, I'm not sure the empty record won't require special handling -- which may complicate the code everywhere we need to examine and deal with the fields of a record.
So I guess my perspective is that complexity is reduced by eliminating special cases, and reflecting that elimination in the type system where feasible.

(I haven't looked at exactly how we handle the empty variant, so maybe it's not as bad as I'm guessing here)

> * I'm not sure about the specific problem that you are trying to fix (disallowing empty records).

Short version: When we have an `assert false` somewhere enforcing a property, by lifting this into the type system, we ensure both that current and future code will obey this, improving maintainability and refactorability, even for those less familiar with the codebase. 

> * I am skeptical about the general idea/approach of trying to shove all invariants into more complex types.

We should judge each individual case to see whether the complexity trade-off is worth it (we're effectively moving complexity from enforcement expressions into the type system).



FWIW: `Core` also has a similar `Nonempty_list` module, but the one here is based on the `Stdlib` list.

**Comment by pitag-ha** [comment]
> Regarding the ppx interface/ast change, we should probably check with ppxlib dev (@pitag-ha maybe ?)

Thank you, @hhugo!

From a PPX ecosystem perspective, changes to strongly used nodes such as `Ppat_record`, `Ptype_record`, `Pexp_record` or `Pcstr_record`,  are really not ideal. Already the change to `Ppat_record` would break about 12 PPXs in our current [ppx-universe](https://github.com/ocaml-ppx/ppx_universe/tree/4.14-bump/rev-deps) (i.e. big opam PPXs work-space that we create for every AST bump to send patch PRs to the opam PPXs we break). And that's assuming that we will absorb the parsetree change in the `ppxlib` API as much as possible and would only break the PPXs, for which we can't absorb it with our current strategies. For the other nodes changed in this PR, the situation would be similar.

However, I also don't want the PPX situation to block the development of the parsetree. So just like I've said on other PRs that would impact us strongly: if you think this change is important, we'll adapt as well as possible. In that case, it's great if you can keep us posted, i.e. ping us again if you decide to go forward with this.

**Comment by gasche** [comment]
(Triaging note: this issue is linked to #11148.)

**Comment by gasche** [comment]
New take on this PR:

- given the feedback from ppxlib folks, it's not clear we want to proceed with the changes to the parsetre
- @bclement-ocp points out that it is a bit awkward to have a user-exposed interface like Parsetree depend on an obscure internal module like Nonempty_list

I would be in favor of giving up. Nonempty_list is cute and using it internally would probably be fine (it could help in some places), but so far there is little left in this PR if we remove the parsetree changes, and I'm not sure about the robustness of forbidding empty records.

**Comment by ferminr** [comment]
> I would be in favor of giving up. Nonempty_list is cute and using it internally would probably be fine 

Regardless of what happens to this PR, I like `Nonempty_list` and I'd be happy if it was promoted from an internal compiler library to `Stdlib` (I can add a wish-list issue myself, or let @smuenzel  do it )

**Comment by gasche** [comment]
If we want to encourage more users to use it, Nonempty_list could be proposed as a self-contained external library. This is much easier than gathering enough support to move into the standard library.

If we want to use it in the compiler, `utils/` sounds right for the time being.

**Comment by gasche** [comment]
With no further feedback on this issue, and the doubts raised, I propose to close and move on for now. If there is interest in reviving the PR, in whole or in pieces, with new comments to defend the approach, I would of course be happy to reopen.
